### DISCUSSION TITLE
turborepo v2 express with nextjs app

### DISCUSSION THREAD

**Discussion by cSarcasme** [vercel/turborepo#8535] [Help]
### Summary

Hi i need little help with turborepo 2

i have create a monorepo with express and pug for backend and for my frontend i use nextjs.
The model i have use for do tat is kitchen-sink.
And i have keep just the express and nextjs part.
Locally no issue when i do turbo run dev my both website work nice.

But when i deploy i have that error, it about the change of pipeline to tasks in the v2 of turborepo and on vercel  have that issue

```
Error: turbo_json_parse_error
  x Found an unknown key `tasks`.
   ,-[4:1]
 4 |     "globalEnv": ["PORT"],
 5 |     "tasks": {
   :     ^^^^^^^
 6 |         "build": {
   `----
Error: Command "turbo run build" exited with 1
```
And this is how is configure my root turbo.json

```
{
	"$schema": "https://turbo.build/schema.json",
	"globalDependencies": ["**/.env.*local"],
	"globalEnv": ["PORT"],
	"tasks": {
		"build": {
			"dependsOn": ["^build"]
		},
		"test": {
			"outputs": ["coverage/**"],
			"dependsOn": []
		},
		"lint": {
			"dependsOn": ["^build"]
		},
		"typecheck": {
			"dependsOn": ["^build"]
		},
		"dev": {
			"dependsOn": ["^build"],
			"cache": false,
			"persistent": true
		},
		"clean": {
			"cache": false
		}
	}
}
```

this the  tubo.json of my frontend folder so nextjs app.
```
{
	"extends": ["//"],
	"tasks": {
		"build": {
			"outputs": [".next/**", "!.next/cache/**"]
		}
	}
}
```

And this the turbo.json of my folder backend where i have express and pug.
```
{
	"extends": ["//"],
	"tasks": {
		"build": {
			"outputs": ["dist/**"]
		}
	}
}
```

I have look to find a solution i have see some people have the same error but they use docker me don't use docker it is on vercel and i don't find a solution for that.

If you need to see other file tell me.

I thank you by advance for any help you will give me thanks.

**Comment by anthonyshew** [comment]
Have you installed `turbo` as a `devDependency` into your root `package.json`?

https://turbo.build/repo/docs/getting-started/installation#repository-installation

**Reply by cSarcasme** [reply]
TY for your answer

i have install the package kitchen sink of the documentation turborepo and it was already install in devdependencies.

I have also install turbo globally because the command turbo was not working when it was a devdependencies so i have both.

I have the version 2 of turbo.

```
Error: turbo_json_parse_error
  x Found an unknown key `tasks`.
   ,-[4:1]
 4 |     "globalEnv": ["PORT"],
 5 |     "tasks": {
   :     ^^^^^^^
 6 |         "build": {
   `----
Error: Command "cd ../.. && turbo run build --filter={apps/frontend}..." exited with 1
```

Also i use npm:

This is my root package.json file

```
{
	"private": true,
	"scripts": {
		"build": "turbo run build",
		"clean": "turbo run clean",
		"dev": "turbo run dev",
		"format": "prettier --write \"**/*.{ts,tsx,md}\"",
		"lint": "turbo run lint",
		"test": "turbo run test",
		"typecheck": "turbo run typecheck"
	},
	"depedencies": {
		"turbo": "^2.0.4"
	},
	"devDependencies": {
		"prettier": "^3.2.5"
	},
	"engines": {
		"node": ">=18"
	},
	"name": "kitchen-sink",
	"packageManager": "npm@10.1.0",
	"workspaces": [
		"apps/*",
		"packages/*"
	]
}
```

My root turbo.json

```
{
	"$schema": "https://turbo.build/schema.json",
	"globalDependencies": ["**/.env.*local"],
	"globalEnv": ["PORT"],
	"tasks": {
		"build": {
			"dependsOn": ["^build"]
		},
		"test": {
			"outputs": ["coverage/**"],
			"dependsOn": []
		},
		"lint": {
			"dependsOn": ["^build"]
		},
		"typecheck": {
			"dependsOn": ["^build"]
		},
		"dev": {
			"dependsOn": ["^build"],
			"cache": false,
			"persistent": true
		},
		"clean": {
			"cache": false
		}
	}
}

```

My turbo.json of my backend folder whos is in apps/backend who work with express and pug.

```

{
	"extends": ["//"],
	"tasks": {
		"build": {
			"outputs": ["dist/**"]
		}
	}
}

```

My turbo.json file in my folder frontend  in apps/frontend who work with nextjs.

```
{
	"extends": ["//"],
	"tasks": {
		"build": {
			"outputs": [".next/**", "!.next/cache/**"]
		}
	}
}
```

When i run the command turbo run dev work perfectly locally but impossible to makes work on deploy on vercel.

If anyone has the solution i am really need it.

**Reply by anthonyshew** [reply]
Is the lockfile up to date?

**Reply by cSarcasme** [reply]
i have fin the solution:

I have find the solution 

First i have done a new instal with the docker example with express and nextjs.

After the root turbo.json need a change because it is always writte pipeline and the v2 of turbo dont accept pipeline, so change pipeline to tasks.

like that:

```
{
    "$schema": "https://turbo.build/schema.json",
    "globalDependencies": ["**/.env.*local"],
    "globalEnv": ["PORT"],
    "tasks": {
        "build": {
            "outputs": ["dist/**", ".next/**", "!.next/cache/**", "public/dist/**"],
            "dependsOn": ["^build"],
            "env": ["NEXT_PUBLIC_API_HOST"]
        },
        "test": {
            "outputs": ["coverage/**"],
            "dependsOn": []
        },
        "lint": {
            "dependsOn": ["^build"]
        },
        "dev": {
            "cache": false,
            "persistent": true
        },
        "clean": {
            "cache": false
        }
    }
}

```

After i ha the same like kitchen sync. and for find the solution i have move turbo from devdepedencies todepedencises like that.

````
{
    "name": "with-docker",
    "version": "0.0.0",
    "private": true,
    "workspaces": [
        "apps/*",
        "packages/*"
    ],
    "scripts": {
        "build": "turbo run build",
        "clean": "turbo run clean",
        "dev": "turbo run dev",
        "format": "prettier --write \"**/*.{ts,tsx,md}\"",
        "lint": "turbo run lint",
        "test": "turbo run test"
    },
    "dependencies": {
        "turbo": "^2.0.4"
    },
    "devDependencies": {
        "@repo/eslint-config": "*",
        "prettier": "^3.2.5"
    },
```

Also before i had use npm i have change to yarn.

And then i have delete roo node modules and lock package.json

And the done a new yarn install


After deploy and finally no error it work


I hope it will help other people i have loose lot of time of that

Have a nice day

**Reply by cSarcasme** [reply]
@anthonyshew

Also i have a other question i dont find how to configure my turborepo file on vercel for the url work like that:

foo.com will be my frontend with next that work

The url for my backend need to be foo.com/admin for example it is my express server with pug and my road api who communicate with the frontend for mongodb.

How i configure my file for do that i dont find a good answer to that.

**Reply by anthonyshew** [reply]
The phrase you're looking for is "multi-zones". This doc can help with that: https://vercel.com/guides/how-can-i-serve-multiple-projects-under-a-single-domain

**Reply by asynchronizd** [reply]
> The phrase you're looking for is "multi-zones". This doc can help with that: https://vercel.com/guides/how-can-i-serve-multiple-projects-under-a-single-domain

this doesn't work with the latest 14.x my dude.

**Reply by anthonyshew** [reply]
I can confirm this works, since this is how `vercel.com` runs today. Here's [a blog post](https://vercel.com/blog/how-vercel-adopted-microfrontends) about us using this technique and [a template](https://vercel.com/templates/next.js/microfrontends) as well.

Hope that helps!
